File: app/src/main/java/me/zhaoweihao/hnuplus/MainActivity.java
package me.zhaoweihao.hnuplus;

import android.graphics.Color;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.view.Window;
import android.widget.ImageView;
import android.widget.TextView;

import butterknife.BindView;
import butterknife.ButterKnife;
import cn.bmob.v3.Bmob;

/**
 * Created by zhaoweihaoChina on 2017/11/9.
 * github:https://github.com/zhaoweihaoChina
 */

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    /**
     * Butter Knife
     */
    @BindView(R.id.hot_layout)  View hotLayout;
    @BindView(R.id.contacts_layout)  View contactsLayout;
    @BindView(R.id.news_layout)  View newsLayout;
    @BindView(R.id.user_layout)  View userLayout;
    @BindView(R.id.hot_image)  ImageView hotImage;
    @BindView(R.id.contacts_image)  ImageView contactsImage;
    @BindView(R.id.news_image)  ImageView newsImage;
    @BindView(R.id.user_image)  ImageView userImage;
    @BindView(R.id.hot_text)  TextView hotText;
    @BindView(R.id.contacts_text)  TextView contactsText;
    @BindView(R.id.news_text)  TextView newsText;
    @BindView(R.id.user_text)  TextView userText;

    /**
     * 用于展示的四个Fragment
     */
    private HotFragment hotFragment;
    private ContactsFragment contactsFragment;
    private NewsFragment newsFragment;
    private UserFragment userFragment;

    /**
     * 用于对Fragment进行管理
     */
    private FragmentManager fragmentManager;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        setContentView(R.layout.activity_main);

        Bmob.initialize(this, "a15e40755375ee7434e6be8c000c184b");

        // 初始化布局元素
        ButterKnife.bind(this);
        bindListener();
        fragmentManager = getSupportFragmentManager();
        // 第一次启动时选中第0个tab
        setTabSelection(0);
    }

    private void bindListener(){
        hotLayout.setOnClickListener(this);
        contactsLayout.setOnClickListener(this);
        newsLayout.setOnClickListener(this);
        userLayout.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.hot_layout:
                // 当点击了消息tab时，选中第1个tab
                setTabSelection(0);
                break;
            case R.id.contacts_layout:
                // 当点击了联系人tab时，选中第2个tab
                setTabSelection(1);
                break;
            case R.id.news_layout:
                // 当点击了动态tab时，选中第3个tab
                setTabSelection(2);
                break;
            case R.id.user_layout:
                // 当点击了设置tab时，选中第4个tab
                setTabSelection(3);
                break;
            default:
                break;
        }
    }

    /**
     * 根据传入的index参数来设置选中的tab页。
     */
    private void setTabSelection(int index) {
        // 每次选中之前先清楚掉上次的选中状态
        clearSelection();
        // 开启一个Fragment事务
        FragmentTransaction transaction = fragmentManager.beginTransaction();
        // 先隐藏掉所有的Fragment，以防止有多个Fragment显示在界面上的情况
        hideFragments(transaction);
        switch (index) {
            case 0:
                // 当点击了消息tab时，改变控件的图片和文字颜色
                hotImage.setImageResource(R.drawable.ic_home_black_24dp);
                hotText.setTextColor(Color.WHITE);
                if (hotFragment == null) {
                    // 如果HotFragment为空，则创建一个并添加到界面上
                    hotFragment = new HotFragment();
                    transaction.add(R.id.content, hotFragment);
                } else {
                    // 如果HotFragment不为空，则直接将它显示出来
                    transaction.show(hotFragment);
                }
                break;
            case 1:
                // 当点击了联系人tab时，改变控件的图片和文字颜色
                contactsImage.setImageResource(R.drawable.ic_home_black_24dp);
                contactsText.setTextColor(Color.WHITE);
                if (contactsFragment == null) {
                    // 如果ContactsFragment为空，则创建一个并添加到界面上
                    contactsFragment = new ContactsFragment();
                    transaction.add(R.id.content, contactsFragment);
                } else {
                    // 如果ContactsFragment不为空，则直接将它显示出来
                    transaction.show(contactsFragment);
                }
                break;
            case 2:
                // 当点击了动态tab时，改变控件的图片和文字颜色
                newsImage.setImageResource(R.drawable.ic_home_black_24dp);
                newsText.setTextColor(Color.WHITE);
                if (newsFragment == null) {
                    // 如果NewsFragment为空，则创建一个并添加到界面上
                    newsFragment = new NewsFragment();
                    transaction.add(R.id.content, newsFragment);
                } else {
                    // 如果NewsFragment不为空，则直接将它显示出来
                    transaction.show(newsFragment);
                }
                break;
            case 3:
            default:
                // 当点击了设置tab时，改变控件的图片和文字颜色
                userImage.setImageResource(R.drawable.ic_home_black_24dp);
                userText.setTextColor(Color.WHITE);
                if (userFragment == null) {
                    // 如果userFragment为空，则创建一个并添加到界面上
                    userFragment = new UserFragment();
                    transaction.add(R.id.content, userFragment);
                } else {
                    // 如果userFragment不为空，则直接将它显示出来
                    transaction.show(userFragment);
                }
                break;
        }
        transaction.commit();
    }

    /**
     * 清除掉所有的选中状态。
     */
    private void clearSelection() {
        hotImage.setImageResource(R.drawable.ic_home_grey_24dp);
        hotText.setTextColor(Color.parseColor("#82858b"));
        contactsImage.setImageResource(R.drawable.ic_home_grey_24dp);
        contactsText.setTextColor(Color.parseColor("#82858b"));
        newsImage.setImageResource(R.drawable.ic_home_grey_24dp);
        newsText.setTextColor(Color.parseColor("#82858b"));
        userImage.setImageResource(R.drawable.ic_home_grey_24dp);
        userText.setTextColor(Color.parseColor("#82858b"));
    }

    /**
     * 将所有的Fragment都置为隐藏状态。
     */
    private void hideFragments(FragmentTransaction transaction) {
        if (hotFragment != null) {
            transaction.hide(hotFragment);
        }
        if (contactsFragment != null) {
            transaction.hide(contactsFragment);
        }
        if (newsFragment != null) {
            transaction.hide(newsFragment);
        }
        if (userFragment != null) {
            transaction.hide(userFragment);
        }
    }

}

File: app/src/main/java/me/zhaoweihao/hnuplus/MainActivity.kt
package me.zhaoweihao.hnuplus

import android.content.Intent
import android.graphics.Color

import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentTransaction
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.Window
import android.widget.Toast
import cn.bmob.v3.Bmob
import cn.bmob.v3.BmobUser
import cn.bmob.v3.exception.BmobException
import cn.bmob.v3.listener.SaveListener

import kotlinx.android.synthetic.main.activity_main.*
import me.zhaoweihao.hnuplus.Constant.Constant
import me.zhaoweihao.hnuplus.Interface.HotInterface
import me.zhaoweihao.hnuplus.JavaBean.MyUser
import me.zhaoweihao.hnuplus.JavaBean.Post
import cn.bmob.v3.listener.UploadFileListener
import cn.bmob.v3.datatype.BmobFile
import org.jetbrains.anko.toast
import java.io.File

import android.app.ProgressDialog


/**
 * Created by zhaoweihaoChina on 2017/11/9.
 * github:https://github.com/zhaoweihaoChina
 */
class MainActivity : AppCompatActivity(){

    /**
     * Four Fragment for display
     */
    private var hotFragment: HotFragment? = null
    private var contactsFragment: ContactsFragment? = null
    private var newsFragment: NewsFragment? = null
    private var userFragment: UserFragment? = null

    /**
     * Used to manage Fragment
     */
    private var fragmentManager: FragmentManager? = null

    /**
     * Used to call fragment method
     */
    private var listener: HotInterface? = null

    private var progressDialog: ProgressDialog? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        setContentView(R.layout.activity_main)

        Bmob.initialize(this, Constant.APP_ID)

        bindListener()
        fragmentManager = supportFragmentManager

        // select 0 tab when first run
        setTabSelection(0)
    }

    private fun bindListener() {
        hot_layout!!.setOnClickListener{setTabSelection(0)}
        contacts_layout!!.setOnClickListener{setTabSelection(1)}
        news_layout!!.setOnClickListener{setTabSelection(2)}
        user_layout!!.setOnClickListener{setTabSelection(3)}
    }

    /**
     * Set the selected tab page based on the index parameter passed in
     */
    private fun setTabSelection(index: Int) {
        // Clear the last selected state before each selection
        clearSelection()
        // Open a Fragment transaction
        val transaction = fragmentManager!!.beginTransaction()
        // Hide all the Fragment first to prevent multiple Fragment displayed on the screen
        hideFragments(transaction)
        when (index) {
            0 -> {
                // When you click the message tab, change the tab's picture and text color
                hot_image!!.setImageResource(R.drawable.ic_home_black_24dp)
                hot_text!!.setTextColor(Color.WHITE)
                if (hotFragment == null) {
                    // If Hot Fragment is empty, create one and add to the screen
                    hotFragment = HotFragment()
                    setListener(hotFragment!!)
                    transaction.add(R.id.content, hotFragment)
                } else {
                    // If Hot Fragment is not empty, it will be displayed directly
                    transaction.show(hotFragment)
                }
            }
            1 -> {
                contacts_image!!.setImageResource(R.drawable.ic_home_black_24dp)
                contacts_text!!.setTextColor(Color.WHITE)
                if (contactsFragment == null) {
                    contactsFragment = ContactsFragment()
                    transaction.add(R.id.content, contactsFragment)
                } else {
                    transaction.show(contactsFragment)
                }
            }
            2 -> {
                news_image!!.setImageResource(R.drawable.ic_home_black_24dp)
                news_text!!.setTextColor(Color.WHITE)
                if (newsFragment == null) {
                    newsFragment = NewsFragment()
                    transaction.add(R.id.content, newsFragment)
                } else {
                    transaction.show(newsFragment)
                }
            }
            3 -> {
                user_image!!.setImageResource(R.drawable.ic_home_black_24dp)
                user_text!!.setTextColor(Color.WHITE)
                if (userFragment == null) {
                    userFragment = UserFragment()
                    transaction.add(R.id.content, userFragment)
                } else {
                    transaction.show(userFragment)
                }
            }
        }
        transaction.commit()
    }

    /**
     * Clear all the selected state
     */
    private fun clearSelection() {
        hot_image!!.setImageResource(R.drawable.ic_home_grey_24dp)
        hot_text!!.setTextColor(Color.parseColor("#82858b"))
        contacts_image!!.setImageResource(R.drawable.ic_home_grey_24dp)
        contacts_text!!.setTextColor(Color.parseColor("#82858b"))
        news_image!!.setImageResource(R.drawable.ic_home_grey_24dp)
        news_text!!.setTextColor(Color.parseColor("#82858b"))
        user_image!!.setImageResource(R.drawable.ic_home_grey_24dp)
        user_text!!.setTextColor(Color.parseColor("#82858b"))
    }

    /**
     * Hide all the Fragments
     */
    private fun hideFragments(transaction: FragmentTransaction) {
        if (hotFragment != null) {
            transaction.hide(hotFragment)
        }
        if (contactsFragment != null) {
            transaction.hide(contactsFragment)
        }
        if (newsFragment != null) {
            transaction.hide(newsFragment)
        }
        if (userFragment != null) {
            transaction.hide(userFragment)
        }
    }

    public fun gotoPostFragment(){
        val intent = Intent(this, PostActivity::class.java)
                startActivityForResult(intent, 1)
    }

    fun setListener(listener: HotInterface) {
        this.listener = listener
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {
            //receive data from PostActivity
            1 -> if (resultCode == RESULT_OK) {
                val returnedData = data!!.getStringExtra("data_return")
                val returnedPath = data!!.getStringExtra("data_return_2")
                //check whether the user selects the picture
                //if returnedPath is null,it means the user did not selects the picture
                if(returnedPath == null){
                    //check the receive data is empty or not
                    if (returnedData == "") {
                        Toast.makeText(this, "empty text", Toast.LENGTH_SHORT).show()
                    } else {
                        //send data to server via bmob sdk
                        val user = BmobUser.getCurrentUser(MyUser::class.java)
                        val post = Post()
                        //show uploading progressdialog
                        progressDialog = ProgressDialog(this)
                        progressDialog!!.setMessage("Uploading...")
                        progressDialog!!.setCancelable(true)
                        progressDialog!!.show()

                        post.content = returnedData
                        post.author = user

                        post.save(object : SaveListener<String>() {

                            override fun done(objectId: String, e: BmobException?) {
                                if (e == null) {
                                    progressDialog!!.hide()
                                    Toast.makeText(this@MainActivity, "post successfully", Toast.LENGTH_SHORT).show()
                                    listener!!.myMethod()
                                } else {
                                    Toast.makeText(this@MainActivity, "post failed", Toast.LENGTH_SHORT).show()
                                }
                            }
                        })

                    }
                }
                else{
                    Log.d("MA",returnedPath)
                    //check the receive data is empty or not
                    if (returnedData == "") {
                        Toast.makeText(this, "empty text", Toast.LENGTH_SHORT).show()
                    } else {
                        //send data to server via bmob sdk
                        val user = BmobUser.getCurrentUser(MyUser::class.java)
                        val post = Post()

                        val bmobFile = BmobFile(File(returnedPath))
                        //show uploading progressdialog
                        progressDialog = ProgressDialog(this)
                        progressDialog!!.setMessage("Uploading...")
                        progressDialog!!.setCancelable(true)
                        progressDialog!!.show()
                        //upload photo
                        bmobFile.uploadblock(object : UploadFileListener() {

                            override fun done(e: BmobException?) {
                                if (e == null) {
                                    progressDialog!!.hide()

                                    post.content = returnedData
                                    post.author = user
                                    post.image = bmobFile
                                    //save post
                                    post.save(object : SaveListener<String>() {

                                        override fun done(objectId: String, e: BmobException?) {
                                            if (e == null) {
                                                Toast.makeText(this@MainActivity, "post successfully", Toast.LENGTH_SHORT).show()
                                                listener!!.myMethod()
                                            } else {
                                                Toast.makeText(this@MainActivity, "post failed", Toast.LENGTH_SHORT).show()
                                            }
                                        }
                                    })

                                } else {
                                    toast("upload file failed : " + e.message)
                                }

                            }

                            override fun onProgress(value: Int?) {
                                // return uploading percentage
                            }
                        })

                    }
                }

            }
                }

    }

}
